src/PlatformAdapter/Native/Exec/NativeExecImpl.cpp

```diff
@@ -0,0 +1,253 @@
+  if (!setup)
+  {
+    for (const std::string& err : errors)
+      fprintf(stderr, "%s: %s\n", args[0].c_str(), err.c_str());
+    return 1;
+  }
+
+  Isolate* isolate = setup->isolate();
+  Environment* env = setup->env();
+
+  // save ptr ref to member filed
+  m_setup = setup.get();
+  m_isolate = isolate;
+  m_env = env;
+  m_loop = setup->event_loop();
+
+  // todo, env hook
+  // if (envHook)
+  // {
+  //   envHook(m_setup->env());
+  // }
+
+  // todo, init script hook
+  // const char* script = getScriptHook ? getScriptHook() : "";
+
+  // todo: inject sdk
+  // link_vgg_sdk_addon(m_setup->env());
+  // const char* requrie_vgg_sdk_script =
+  //   "const vggSdkAddon = process._linkedBinding('vgg_sdk_addon');"
+  //   "globalThis.vggSdk = new vggSdkAddon.VggSdk()";
+
+  {
+    Locker locker(isolate);
+    Isolate::Scope isolate_scope(isolate);
+    HandleScope handle_scope(isolate);
+    Context::Scope context_scope(setup->context());
+
+    MaybeLocal<Value> loadenv_ret = node::LoadEnvironment(
+      env,
+      "globalThis.require = require('module').createRequire(process.cwd() + '/');");
+
+    if (loadenv_ret.IsEmpty()) // There has been a JS exception.
+      return 1;
+
+    if (m_state != CANCELLED)
+    {
+      const std::lock_guard<std::mutex> lock(m_state_mutex);
+      m_state = RUNNING;
+
+      INFO("#exec, init keep alive timer");
+      uv_timer_init(setup->event_loop(), &m_keep_alive_timer);
+      uv_timer_start(
+        &m_keep_alive_timer,
+        [](uv_timer_t* timer) {},
+        std::numeric_limits<uint64_t>::max(),
+        0);
+    }
+
+    INFO("#exec, node run");
+    exit_code = node::SpinEventLoop(env).FromMaybe(1);
+
+    node::Stop(env);
+  }
+
+  m_state = DEAD;
+  INFO("#exec, node exit");
+
+  return exit_code;
+}
+
+void NativeExecImpl::erase_task(NativeEvalTask* task)
+{
+  const std::lock_guard<std::mutex> lock(m_tasks_mutex);
+  m_tasks.erase(task);
+}
+
+void NativeExecImpl::stop_node_thread_safe()
+{
+  if (m_state != RUNNING)
+  {
+    const std::lock_guard<std::mutex> lock(m_state_mutex);
+    m_state = CANCELLED;
+
+    INFO("#exec, cancel node");
+    return;
+  }
+
+  uv_async_init(m_loop,
+                &m_stop_timer_async,
+                [](uv_async_t* async)
+                {
+                  DEBUG("#exec, stop_node_thread_safe, stop timer");
+                  auto self_ptr = static_cast<NativeExecImpl*>(async->data);
+                  uv_timer_stop(&self_ptr->m_keep_alive_timer);
+                  uv_close(reinterpret_cast<uv_handle_t*>(&self_ptr->m_keep_alive_timer), nullptr);
+                  uv_close(reinterpret_cast<uv_handle_t*>(&self_ptr->m_stop_timer_async), nullptr);
+                });
+  m_stop_timer_async.data = static_cast<void*>(this);
+  uv_async_send(&m_stop_timer_async);
+}
+
+void NativeExecImpl::run_task(NativeEvalTask* task)
+{
+  DEBUG("#evalScript, before eval");
+  eval(task->m_code);
```

Comment:
should we care about the return code of `eval`? `schedual_eval` always returns true, so we cannot get the actual eval result.

Reply:
add log to print the ret code, and there are logs in the `eval` method.